flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/join/TimeBoundedStreamInnerJoin.scala

```diff
@@ -131,340 +116,308 @@ class TimeBoundedStreamInnerJoin(
+              i += 1
+            }
+          } else {
+            while (i < oppositeRows.size) {
+              joinFunction.join(oppositeRows.get(i), row, cRowWrapper)
+              i += 1
             }
-          } else if (oppositeTime > oppositeUpperBound) {
-            //TODO If the keys are ordered, can we break here?
           }
         }
-      } else {
-        //TODO Need some extra logic here?
-        LOG.warn(s"$record is out-of-date.")
+        // We could do the short-cutting optimization here once we get a state with ordered keys.
       }
     }
+    // We need to deal with the late data in the future.
   }
 
   /**
-    * Register a timer for cleaning up records in a specified time.
+    * Register a timer for cleaning up rows in a specified time.
     *
     * @param ctx               the context to register timer
-    * @param timeForRecord     time for the input record
+    * @param rowTime           time for the input row
     * @param oppositeWatermark watermark of the opposite stream
     * @param timerState        stores the timestamp for the next timer
-    * @param leftRecord        record from the left or the right stream
+    * @param leftRow           whether this row comes from the left stream
     * @param firstTimer        whether this is the first timer
     */
   private def registerCleanUpTimer(
-    ctx: CoProcessFunction[CRow, CRow, CRow]#Context,
-    timeForRecord: Long,
-    oppositeWatermark: Long,
-    timerState: ValueState[Long],
-    leftRecord: Boolean,
-    firstTimer: Boolean): Unit = {
-    val cleanUpTime = timeForRecord + (if (leftRecord) leftRelativeSize else rightRelativeSize) +
-      allowedLateness + 1
-    registerTimer(ctx, !leftRecord, cleanUpTime)
-    LOG.debug(s"Register a clean up timer on the ${if (leftRecord) "RIGHT" else "LEFT"} state:"
-      + s" timeForRecord = ${timeForRecord}, cleanUpTime = ${cleanUpTime}, oppositeWatermark = " +
-      s"${oppositeWatermark}")
-    timerState.update(cleanUpTime)
-    if (cleanUpTime <= oppositeWatermark + allowedLateness && firstTimer) {
-      backPressureSuggestion =
-        if (leftRecord) (oppositeWatermark + allowedLateness - cleanUpTime)
-        else -(oppositeWatermark + allowedLateness - cleanUpTime)
-      LOG.warn("The clean timer for the " +
-        s"${if (leftRecord) "left" else "right"}" +
-        s" stream is lower than ${if (leftRecord) "right" else "left"} watermark." +
-        s" requiredTime = ${formatTime(cleanUpTime)}, watermark = ${formatTime(oppositeWatermark)},"
-        + s"backPressureSuggestion = " + s"${backPressureSuggestion}.")
+      ctx: CoProcessFunction[CRow, CRow, CRow]#Context,
+      rowTime: Long,
+      oppositeWatermark: Long,
+      timerState: ValueState[Long],
+      leftRow: Boolean,
+      firstTimer: Boolean): Unit = {
+    val cleanupTime = if (leftRow) {
+      rowTime + leftRelativeSize + cleanupDelay + allowedLateness + 1
```

Comment:
Thanks for this suggestion and I understand your means here. 

IMO, as long as the stream is continuous, the timer should always be registered in the `removeExpiredRows()`, except for the first one. Thus in real applications, I think there's not so much difference whether we apply the `cleanupDelay` in `removeExpiedRows()` or in `registerCleanUpTimer()`. Moreover, calculating cleanup time differently may confuse the developers/users and it will also make the test complicated (i.e., not easy to decide when a row should be removed). What do you think?

Reply:
we can also leave it here. It doesn't affect the semantics of the join and should have not much effect if you eagerly clean the state in the `processElement()` methods.